### PR TITLE
Correctly parse ado without any statements in it

### PR DESCRIPTION
**Pull request by MonoidMusician**
Fixes https://github.com/natefaubion/purescript-tidy/issues/108

The issue was that the parser wasn't backtracking enough to abort `valueParser` and fall back to `pure []`, so instead I parse the layout tokens up front so we can commit to the right path early. Perhaps it could be more elegant.

Tested on the examples in the issue and on our work codebase. Do you want me to add a test to `test/Main.purs`?

### CODE REVIEW

File: src/PureScript/CST/Parser.purs [hunk 1]
```diff
@@ -97,9 +97,11 @@ layoutNonEmpty valueParser = ado
 
 layout :: forall a. Parser a -> Parser (Array a)
 layout valueParser =
-  tokLayoutStart *> values <* tokLayoutEnd
+  tokLayoutStart *> values
   where
-  values = (go =<< valueParser) <|> pure []
```
Comment (natefaubion) [inline src/PureScript/CST/Parser.purs:102]: This seems like it should be equivalent to `optional`? If we had an empty layout, then valueParser should immediately fail if given a TokLayoutEnd, shouldn't it? It's not obvious to me what "backtracking far enough" means then, and that there isn't some other internal bug. Do we know where backtracking is halting?

**Comment by natefaubion** [comment]
I think that this should have a regression test, yes. We test against publicly available code, but this was triggered by an edge case that doesn't exist in any public code.

File: src/PureScript/CST/Parser.purs (hunk 1 above)
Comment (MonoidMusician) [reply src/PureScript/CST/Parser.purs:102]: Okay, I was a little hasty in ascribing it to backtracking. Putting `try` in there does not help.

What's really messing it up is the indentation-based parser recovery when we call it in `parseAdo` (this is the only usage of `layout`). An an example change, this manages to parse the empty `ado in`:

```diff
--- a/src/PureScript/CST/Parser.purs
+++ b/src/PureScript/CST/Parser.purs
@@ -673,7 +673,7 @@ parseDo = do
 parseAdo :: Parser (Recovered Expr)
 parseAdo = do
   keyword <- tokQualifiedKeyword "ado"
-  statements <- layout (recoverDoStatement parseDoStatement)
+  statements <- layout parseDoStatement
   in_ <- tokKeyword "in"
   result <- parseExpr
   pure $ ExprAdo { keyword, statements, in: in_, result }
```

Comment (natefaubion) [reply src/PureScript/CST/Parser.purs:102]: It makes sense to me that there is something odd with recovery, but I'm not sure what. In the case of `ado in`, I would expect `recoverIndent` to immediately hit a `TokLayoutEnd` where `col == indent`, resulting in empty `tokens`, yielding `Nothing`, which would propagate the error.
https://github.com/natefaubion/purescript-language-cst-parser/blob/5afff30ddafc84317d35a4d091a00563d818b899/src/PureScript/CST/Parser.purs#L1155-L1164
https://github.com/natefaubion/purescript-language-cst-parser/blob/5afff30ddafc84317d35a4d091a00563d818b899/src/PureScript/CST/Parser/Monad.purs#L149-L151
Maybe something is wrong with the state threading in `recover`.

Comment (MonoidMusician) [reply src/PureScript/CST/Parser.purs:102]: Okay, I'll push a test for this case and dig deeper.

File: src/PureScript/CST/Parser.purs [hunk 2]
```diff
@@ -673,7 +665,20 @@ parseDo = do
...
+    let
+      -- `recoverDoStatement` recovers too much if it is immediately
+      -- confronted with `TokLayoutEnd`, since that is associated with a
+      -- `layoutStack` _of the parent_ as opposed to the stuff we actually
+      -- want to recover, which we would correctly guess if we saw a statement
+      -- or two inside the block
+      valueParser = recoverDoStatement parseDoStatement
+      nonEmptyCase =
+        Array.cons <$> valueParser <*> many (tokLayoutSep *> valueParser)
+    void tokLayoutStart
```
Comment (natefaubion) [inline src/PureScript/CST/Parser.purs:679]:
```suggestion
    _ <- tokLayoutStart
```

**Comment by natefaubion** [comment]
Thanks!
